relayer/middleware/missedVaas/missedVaas.middleware.ts

```diff
@@ -0,0 +1,443 @@
+import * as grpcWebNodeHttpTransport from "@improbable-eng/grpc-web-node-http-transport";
+import { Middleware } from "../../compose.middleware";
+import { Context } from "../../context";
+import Redis, { Cluster, RedisOptions } from "ioredis";
+import { ChainId, getSignedVAAWithRetry } from "@certusone/wormhole-sdk";
+import {
+  defaultWormholeRpcs,
+  ParsedVaaWithBytes,
+  RelayerApp,
+  RelayerEvents,
+} from "../../application";
+import { Logger } from "winston";
+import { createPool, Pool } from "generic-pool";
+import { dbg, minute, sleep } from "../../utils";
+import { RedisConnectionOpts } from "../../storage/redis-storage";
+import { GetSignedVAAResponse } from "@certusone/wormhole-sdk-proto-web/lib/cjs/publicrpc/v1/publicrpc";
+
+const IN_PROGRESS_TIMEOUT = 5 * minute;
+
+export type { RedisOptions };
+export interface MissedVaaOpts extends RedisConnectionOpts {
+  checkForMissedVaasEveryMs?: number;
+  wormholeRpcs?: string[];
+  logger?: Logger;
+}
+
+export interface VaaKey {
+  emitterChain: number;
+  emitterAddress: string; // todo: string, wh format
+  seq: bigint;
+}
+
+type FetchVaaFn = (vaa: VaaKey) => Promise<GetSignedVAAResponse>;
+type ProcessVaaFn = (x: Buffer) => Promise<void>;
+type TryFetchAndProcessFn = (
+  redis: Redis | Cluster,
+  vaaKey: VaaKey,
+  logger?: Logger,
+) => Promise<boolean>;
+
+/**
+ * Storage schema
+ * chain/emitter -> sortedSet -> seq
+ *
+ * Job:
+ * - requery for missed vaas
+ * - query next vaa
+ *
+ * Middleware
+ * - requery for missed vaas since last seen (not all)
+ */
+
+export function missedVaas(
+  app: RelayerApp<any>,
+  opts: MissedVaaOpts,
+): Middleware {
+  // set defaults
+  opts.redis = opts.redis || { host: "localhost", port: 6379 };
+  opts.redis.keyPrefix = opts.namespace;
+  opts.wormholeRpcs = opts.wormholeRpcs ?? defaultWormholeRpcs[app.env];
+
+  const redisPool = createRedisPool(opts);
+
+  // mark vaa processed when app emits "Added" event
+  app.addListener(RelayerEvents.Added, (vaa: ParsedVaaWithBytes) => {
+    redisPool.use(redis =>
+      markProcessed(
+        redis,
+        {
+          emitterAddress: vaa.emitterAddress.toString("hex"),
+          emitterChain: vaa.emitterChain,
+          seq: vaa.sequence,
+        },
+        opts.logger,
+      ),
+    );
+  });
+
+  // construct dependency
+  const fetchVaaFn = (vaaKey: VaaKey) => fetchVaa(opts.wormholeRpcs, vaaKey);
+
+  // start worker
+  setTimeout(() => startMissedVaaWorker(redisPool, app, fetchVaaFn, opts), 100); // start worker once config is done.
+
+  // return middleware
+  return middlewareFn(opts, fetchVaaFn, redisPool);
+}
+
+function middlewareFn(
+  opts: MissedVaaOpts,
+  fetchVaa: FetchVaaFn,
+  redisPool: Pool<Redis | Cluster>,
+): Middleware {
+  return async (ctx: Context, next) => {
+    const wormholeRpcs = opts.wormholeRpcs ?? defaultWormholeRpcs[ctx.env];
+
+    let vaa = ctx.vaa;
+    if (!vaa) {
+      await next();
+      return;
+    }
+
+    redisPool.use(async redis => {
```

Comment:
Wouldn't it make sense to just remove this part and let the worker started by `startMissedVaaWorker` handle this cases too? I think the algorithm of `missedVaaJob` would catch this cases just fine and we would gain two things:
- all the missed-vaa detection/processing happens in only one place
- there's a little bit less of coupling between different VAAs (for example a VAA might be delayed if previous VAAs were missing)
I'm assumming that since the new missed vaa detection looks for gaps between the processed VAAs then the order of VAAs shouldn't be a problem

Reply:
True! Removed the middleware entirely